### PR TITLE
WebView implementation of the wallet functionalities

### PR DESCRIPTION
**Pull request by ishrak-imam**
Because of not having the WASM support in react-native environment we were using [react-native-substrate-sign](https://github.com/paritytech/react-native-substrate-sign) for wallet related functionalities. But that project is not maintained anymore and it also doesn't support the address formatting of the parachain networks.

That's why we need to bring back one of the other [options](https://github.com/litentry/litentry-app/pull/553) that was considered before, when we faced this performance issue while using polkadot keyring.

The idea is to load the polkadot keyring related packages inside a WebView and interact with that from the app.

Following are the functionalities that we currently need 

- [x] Generate mnemonic
- [x] Create account
- [x] Add external account
- [x] Remove account
- [x] Import account from seed
- [x] Import account from JSON
- [x] Export account to JSON file
- [x] Sign transactions
- [x] Test on real devices 
- [x] Clean up the code

### CODE REVIEW

File: src/ui/components/Account/AddExternalAccount.tsx [hunk 1]
```diff
@@ -78,11 +81,18 @@ export function AddExternalAccount({onClose}: Props) {
...
+      //   meta: {name: '', network: currentNetwork.key, isFavorite: false},
+      //   isExternal: true,
+      // });
+
+      addExternalAccount({
         address: state.address,
-        meta: {name: '', network: currentNetwork.key, isFavorite: false},
-        isExternal: true,
+        network: currentNetwork.key,
+        isFavorite: false,
```
Comment (silva-fj) [inline src/ui/components/Account/AddExternalAccount.tsx]: Let's make this argument optional with default to `false`

File: src/ui/screens/AccountsScreen.tsx [hunk 2]
```diff
@@ -14,6 +13,8 @@ import {Account} from '@ui/components/Account/Account';
 import {Padder} from '@ui/components/Padder';
 import {AccountsGuide} from '@ui/components/Account/AccountsGuide';
 import {AddExternalAccount} from '@ui/components/Account/AddExternalAccount';
+import type {Account as AppAccount} from '@polkadotApi/types';
```
Comment (silva-fj) [inline src/ui/screens/AccountsScreen.tsx:16]: What's the difference between `Account` and `AppAccount`? Why not just Account?

File: polkadot-api/src/action.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
+type WebViewActionPayload = Record<string, unknown>;
+
+export enum ACTION_TYPES {
```
Comment (silva-fj) [inline polkadot-api/src/action.ts]:
```suggestion
export enum ACTION_TYPE {
```

It should be singular, it represents a single value of multiple variants

File: polkadot-api/src/polkadotApi.ts [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+        window.ReactNativeWebView.postMessage(
+          JSON.stringify({
+            type: forgetAccount.resultType,
+            payload: {address: payload.address},
+          }),
+        );
+        break;
+      }
+
+      // case 'TOGGLE_FAVORITE': {
```
Comment (silva-fj) [inline polkadot-api/src/polkadotApi.ts]: What about this?

File: src/polkadotApi/PolkadotApiWebView.tsx [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+            webViewRef.current?.postMessage(exportAccount.getAction(payload));
+          }),
+        verifyCredentials: (credentials: SignCredentials) =>
+          new Promise((resolve) => {
+            resolversRef.current.resolveVerifyCredentials = resolve;
+            webViewRef.current?.postMessage(verifyCredentials.getAction(credentials));
+          }),
+        sign: (message: string, credentials: SignCredentials) =>
+          new Promise((resolve, reject) => {
+            console.log(message, credentials);
```
Comment (silva-fj) [inline src/polkadotApi/PolkadotApiWebView.tsx]: Let's remove this

File: src/polkadotApi/PolkadotApiWebView.tsx [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+            resolversRef.current.exportAccountPromise.resolve = resolve;
+            resolversRef.current.exportAccountPromise.reject = reject;
+            webViewRef.current?.postMessage(exportAccount.getAction(payload));
+          }),
+        verifyCredentials: (credentials: SignCredentials) =>
+          new Promise((resolve) => {
+            resolversRef.current.resolveVerifyCredentials = resolve;
+            webViewRef.current?.postMessage(verifyCredentials.getAction(credentials));
+          }),
+        sign: (message: string, credentials: SignCredentials) =>
```
Comment (silva-fj) [inline src/polkadotApi/PolkadotApiWebView.tsx:187]: Let's add the return type here so we don't have to do type assertion in the `TxContext`

Comment (silva-fj) [reply src/polkadotApi/PolkadotApiWebView.tsx:187]:
```
Promise<{signed: `0x${string}`}>

File: src/ui/screens/AccountsScreen.tsx (hunk 2 above)
Comment (ishrak-imam) [reply src/ui/screens/AccountsScreen.tsx:16]: We already have an Account import (component) in this file. That's why needed to rename the type.

File: polkadot-api/src/polkadotApi.ts (hunk 4 above)
Comment (ishrak-imam) [reply polkadot-api/src/polkadotApi.ts]: This was a less priority functionality, updating the account meta data to mark favourite accounts. I have already added this in a separate branch. Will open a separate PR once the clean-up is merged into this.

File: src/polkadotApi/PolkadotApiWebView.tsx (hunk 5 above)
Comment (ishrak-imam) [reply src/polkadotApi/PolkadotApiWebView.tsx]: Did this in the clean-up.

File: src/context/TxContext.tsx [hunk 7]
```diff
@@ -50,9 +50,11 @@ export type StartConfig = {
 export function TxProvider({children}: TxProviderProps): React.ReactElement {
   const {BottomSheet, openBottomSheet, closeBottomSheet} = useBottomSheet();
   const signTransactionRef = useRef<(value: SignerResult) => void>();
-  const showPreviewRef = useRef<(txPayload: SignerPayloadJSON, seed?: string) => Promise<void>>();
+  const showPreviewRef = useRef<(txPayload: SignerPayloadJSON, credentials?: SignCredentials) => Promise<void>>();
+  // const showPreviewRef = useRef<(txPayload: SignerPayloadJSON, pair?: KeyPair?) => Promise<void>>();
```
Comment (silva-fj) [inline src/context/TxContext.tsx]: Let's remove this as well.

File: src/ui/screens/ImportAccountWithJsonFileScreen.tsx [hunk 8]
```diff
@@ -24,28 +23,22 @@ export function ImportAccountWithJsonFileScreen({navigation}: {navigation: Navig
...
-        };
-        addAccount(newAcc);
-        SecureKeychain.setPasswordByServiceId(password, 'BIOMETRICS', parsedJson.address);
-
+        const account = await restoreAccount({
+          json: parsedJson,
+          password,
+          network: currentNetwork.key,
+        });
+        SecureKeychain.setPasswordByServiceId(password, 'BIOMETRICS', account.address as string);
```
Comment (silva-fj) [inline src/ui/screens/ImportAccountWithJsonFileScreen.tsx:37]: What is the type of `account.address` ?

File: src/ui/screens/MyAccountScreen.tsx [hunk 9]
```diff
@@ -37,8 +39,13 @@ type ScreenProps = {
 export function MyAccountScreen({navigation, route}: ScreenProps) {
   const {address} = route.params;
   const {data: accountInfo} = useAccount(address);
-  const {accounts, removeAccount} = useAccounts();
+
+  // const {accounts, removeAccount} = useAccounts();
```
Comment (silva-fj) [inline src/ui/screens/MyAccountScreen.tsx]: Let's clean this as well

File: src/ui/screens/MyAccountScreen.tsx [hunk 10]
```diff
@@ -132,7 +139,8 @@ export function MyAccountScreen({navigation, route}: ScreenProps) {
                 {
                   text: 'Delete',
                   onPress: () => {
-                    removeAccount(address);
+                    // removeAccount(address);
```
Comment (silva-fj) [inline src/ui/screens/MyAccountScreen.tsx]: This too

**Review by silva-fj — COMMENTED** [review]
Awesome! 👍🏼

File: src/ui/screens/ImportAccountWithJsonFileScreen.tsx (hunk 8 above)
Comment (ishrak-imam) [reply src/ui/screens/ImportAccountWithJsonFileScreen.tsx:37]: unknown, this is coming from the WebView.

**Review by silva-fj — APPROVED** [review]
Let's get this merged

**Pull request merged**